consensus/parlia/parlia.go

```diff
@@ -1229,6 +1669,79 @@ func (p *Parlia) applyTransaction(
 	return nil
 }
 
+func (p *Parlia) SetVotePool(votePool consensus.VotePool) {
+	p.votePool = votePool
+}
+
+func (p *Parlia) GetHighestJustifiedHeader(chain consensus.ChainHeaderReader, header *types.Header) *types.Header {
+	if chain == nil || header == nil {
+		return nil
+	}
+
+	// There won't any attestation in first two blocks, return root.
+	if header.Number.Uint64() < 2 {
+		return chain.GetHeaderByNumber(0)
+	}
+
+	parent := chain.GetHeaderByHash(header.ParentHash)
+	if parent == nil {
+		return nil
+	}
+	snap, err := p.snapshot(chain, parent.Number.Uint64(), parent.Hash(), nil)
+	if err != nil {
+		log.Error("Unexpected error when getting snapshot",
+			"error", err, "blockNumber", parent.Number.Uint64(), "blockHash", parent.Hash())
+		return nil
+	}
+
+	// If there is no vote justified block, then return root.
+	if snap.Attestation == nil {
+		return chain.GetHeaderByNumber(0)
+	}
+	// Return naturally justified block.
+	if snap.Number-snap.Attestation.TargetNumber > naturallyJustifiedDist {
+		return FindAncientHeader(parent, naturallyJustifiedDist, chain, nil)
```

Comment:
Is the edge case correct? The `naturallyJustifiedDist` block before `parent`, not current header?

Reply:
I mistake it, it should be the current header